File: Data Structures & Algorithms/4sum/submission-3.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        res = set()

        left, right = 0, len(nums) - 1
        count = 0
        for left in range(len(nums) - 1):
            for right in range(len(nums) - 1, left, -1):
                l, r = left + 1, right - 1
                curr = nums[left] + nums[right]
                while l < r:
                    tmp = curr + nums[l] + nums[r]
                    if tmp == target:
                        res.add((nums[left], nums[l], nums[r], nums[right]))
                        r -= 1
                    elif tmp < target:
                        l += 1
                    else:
                        r -= 1

        return [[a, b, c, d] for a, b, c, d in res]

File: Data Structures & Algorithms/sqrtx/submission-3.cpp
class Solution {
public:
    int mySqrt(int x) {
        if (x < 2)
        {
            return x;
        }

        int left(1);
        long long curr(0), cnext(0);
        int right = x / 2;
        long long mid = left + (right - left) / 2;
        while (left < right)
        {
            curr = mid * mid;
            if (curr == x)
            {
                return mid;
            }
            cnext = (mid + 1) * (mid + 1);
            if (cnext == x)
            {
                return mid + 1;
            }
            if (curr < x)
            {
                if (cnext > x)
                {
                    return mid;
                }
                left = mid + 1;
            }
            else
            {
                right = mid - 1;
            }
            mid = left + (right - left) / 2;
        }
        return mid;
    }
};